common/meta/raft-store/src/state_machine/sm.rs

```diff
@@ -658,6 +674,88 @@ impl StateMachine {
         Ok(Some(seq_kv_value))
     }
 
+    fn txn_incr_seq(&self, key: &str, txn_tree: &TransactionSledTree) -> u64 {
+        let seq_sub_tree = txn_tree.key_space::<Sequences>();
+        let old: Option<SeqNum> = seq_sub_tree.get(&key.to_string()).unwrap();
+        let curr = old.unwrap_or_default() + 1;
+        seq_sub_tree.insert(&key.to_string(), &curr).unwrap();
+
+        tracing::debug!("applied IncrSeq: {}={}", key, curr);
+
+        curr.0
+    }
+
+    fn sub_txn_tree_upsert<'s, V, KS>(
+        &'s self,
+        sub_tree: &AsTxnKeySpace<'s, KS>,
+        txn_tree: &TransactionSledTree,
+        key: &KS::K,
+        seq: &MatchSeq,
+        value_op: Operation<V>,
+        value_meta: Option<KVMeta>,
+    ) -> (Option<SeqV<V>>, Option<SeqV<V>>)
+    where
+        V: Clone + Debug,
+        KS: SledKeySpace<V = SeqV<V>>,
+    {
+        let prev = sub_tree.get(key).unwrap();
+
+        // If prev is timed out, treat it as a None.
+        let prev = Self::unexpired_opt(prev);
+
+        if seq.match_seq(&prev).is_err() {
+            return (prev.clone(), prev);
+        }
+
+        // result is the state after applying an operation.
+        let result = self.sub_txn_tree_do_update(
+            sub_tree,
+            txn_tree,
+            key,
+            prev.clone(),
+            value_meta,
+            value_op,
+        );
+
+        tracing::debug!("applied upsert: {} {:?}", key, result);
+        (prev, result)
+    }
+
+    /// Update a record into a sled tree sub tree, defined by a KeySpace, without seq check.
+    ///
+    /// TODO(xp); this should be a method of sled sub tree
+    fn sub_txn_tree_do_update<'s, V, KS>(
+        &'s self,
+        sub_tree: &AsTxnKeySpace<'s, KS>,
+        txn_tree: &TransactionSledTree,
```

Comment:
There is a reference to `TransactionSledTree` in the `sub_tree`.
Thus it seems `txn_tree` is not necessary?

Reply:
Yup, I'm working on it